transmission_interface/src/transmission_parser.cpp

```diff
@@ -12,68 +12,250 @@
+        g_logger, "No type element found in transmission '" <<
+          transmission.name << "'.");
+      throw std::runtime_error("No type element found in transmission");
+    }
+    if (!type_child->GetText()) {
+      RCLCPP_ERROR_STREAM(
+        g_logger, "Skipping empty type element in transmission '" <<
+          transmission.name << "'.");
+      throw std::runtime_error("empty type element in transmission");
+    }
+    transmission.control_type = type_child->GetText();
+
+    // Load joints
+    if (!parseJoints(trans_it, transmission.joints)) {
+      RCLCPP_ERROR_STREAM(
+        g_logger, "Failed to load joints for transmission '" <<
+          transmission.name << "'.");
+      throw std::runtime_error("Failed to load joints for transmission");
+    }
+
+    // Load actuators
+    if (!parseActuators(trans_it, transmission.actuators)) {
+      RCLCPP_ERROR_STREAM(
+        g_logger, "Failed to load actuators for transmission '" <<
+          transmission.name << "'.");
+      throw std::runtime_error("Failed to load actuators for transmission");
+    }
+
+    // Save loaded transmission
+    transmissions.push_back(transmission);
+  }  // end for <transmission>
+
+  if (transmissions.empty()) {
+    RCLCPP_DEBUG_STREAM(g_logger, "No valid transmissions found.");
+  }
+
+  return true;
+}
+
+bool TransmissionParser::parseJoints(TiXmlElement * trans_it, std::vector<JointInfo> & joints)
+{
+  // Loop through each available joint
+  TiXmlElement * joint_it = nullptr;
+  for (joint_it = trans_it->FirstChildElement("joint"); joint_it;
+    joint_it = joint_it->NextSiblingElement("joint"))
+  {
+    // Create new joint
+    transmission_interface::JointInfo joint;
+
     // Joint name
-    auto joint_it = trans_it->FirstChildElement(kJointTag);
-    if (!joint_it) {
-      throw std::runtime_error("no joint child element found");
+    if (joint_it->Attribute("name")) {
+      joint.name = joint_it->Attribute("name");
+      if (joint.name.empty()) {
+        RCLCPP_ERROR_STREAM(g_logger, "Empty name attribute specified for joint.");
+        continue;
+      }
+    } else {
+      RCLCPP_ERROR_STREAM(g_logger, "No name attribute specified for joint.");
+      return false;
     }
 
-    const std::string joint_name = joint_it->Attribute("name");
-    if (joint_name.empty()) {
-      throw std::runtime_error("no joint name attribute set");
+    TiXmlElement * role_it = joint_it->FirstChildElement("role");
+    if (role_it) {
+      joint.role_ = role_it->GetText() ? role_it->GetText() : std::string();
     }
 
-    const auto hardware_interface_it = joint_it->FirstChildElement(kHardwareInterfaceTag);
-    if (!hardware_interface_it) {
-      throw std::runtime_error(
-              "no hardware interface tag found under transmission joint" + joint_name);
+    // Hardware interfaces (required)
+    TiXmlElement * hw_iface_it = nullptr;
+    for (hw_iface_it = joint_it->FirstChildElement("hardwareInterface"); hw_iface_it;
```

Comment:
We might want to change the xml tags later (I hope to get rid of CamelCase tags) and would thus propose to put the strings for the tags in a constant.

Reply:
@guru-florida check the example solution [here](https://github.com/ros-controls/ros2_control/blob/master/hardware_interface/src/component_parser.cpp)